include/message_filters/sync_policies/approximate_epsilon_time.h

```diff
@@ -0,0 +1,357 @@
+    }
+    const auto & events_of_this_type = std::get<Is>(events_);
+    if (0u == events_of_this_type.size()) {
+      // this condition should not happen
+      return current;
+    }
+    auto candidate = mt::TimeStamp<typename ThisEventType::Message>::value(*events_of_this_type.at(0).getMessage());
+    if (current.first > candidate) {
+      return std::make_pair(candidate, Is);
+    }
+    return current;
+  }
+
+  template <size_t ... Is>
+  TimeIndexPair
+  get_older_timestamp_helper(std::index_sequence<Is...> const &)
+  {
+    TimeIndexPair older{
+      rclcpp::Time(std::numeric_limits<int64_t>::max(), RCL_ROS_TIME), std::numeric_limits<size_t>::max()};
+    ((older = get_older_timestamp_between<Is>(older)), ...);
+    return older;
+  }
+
+  TimeIndexPair
+  get_older_timestamp()
+  {
+    return get_older_timestamp_helper(std::make_index_sequence<9u>());
+  }
+
+  template <size_t Is>
+  bool
+  check_timestamp_within_epsilon(const TimeIndexPair & older)
+  {
+    namespace mt = message_filters::message_traits;
+    using ThisEventType = typename std::tuple_element<Is, Events>::type;
+    if constexpr (Is >= RealTypeCount::value) {
+      return true;
+    }
+    if (Is == older.second) {
+      return true;
+    }
+    const auto & events_of_this_type = std::get<Is>(events_);
+    if (0u == events_of_this_type.size()) {
+      // this condition should not happen
+      return false;
+    }
+    auto ts = mt::TimeStamp<typename ThisEventType::Message>::value(*events_of_this_type.at(0).getMessage());
+    if (older.first + epsilon_ >= ts) {
+      return true;
+    }
+    return false;
+  }
+
+  template <size_t ... Is>
+  bool
+  check_all_timestamp_within_epsilon_helper(
+    const TimeIndexPair & older, std::index_sequence<Is...> const &)
+  {
+    bool valid = true;
+    ((valid &= check_timestamp_within_epsilon<Is>(older)), ...);
+    return valid;
+  }
+
+  bool
+  check_all_timestamp_within_epsilon(const TimeIndexPair & older)
+  {
+    return check_all_timestamp_within_epsilon_helper(
+      older, std::make_index_sequence<9u>());
+  }
+
+  // void erase_old_events()
+  // {
+  //   auto old = get_older_timestamp();
+  //   erase_old_events_within_ts(old);
+  // }
+
+  template<size_t Is>
+  void
+  erase_begin_of_vector()
+  {
+    if constexpr (Is >= RealTypeCount::value) {
+      return;
+    }
+    auto & this_vector = std::get<Is>(events_);
+    if (this_vector.begin() != this_vector.end()) {
+      this_vector.erase(this_vector.begin());
+      if (this_vector.empty()) {
+        --number_of_non_empty_events_;
+      }
+    }
+  }
+
+  template <size_t ... Is>
+  void
+  erase_begin_of_vectors_helper(std::index_sequence<Is...> const &)
+  {
+    ((erase_begin_of_vector<Is>()), ...);
+  }
+
+  void erase_begin_of_vectors()
+  {
+    return erase_begin_of_vectors_helper(std::make_index_sequence<9u>());
+  }
+
+  template<size_t Is>
+  void
+  erase_begin_of_vector_if_within_ts(rclcpp::Time timestamp)
```

Comment:
```suggestion
  erase_beginning_of_vector_if_earlier_than_ts(rclcpp::Time timestamp)
```

Reply:
it's removed if the event happened between `timestamp` and `timestamp+epsilon`, i.e. it's the event happened approximately on sync with ts.
Maybe `erase_beginning_of_vector_if_on_sync_with_ts()` is a better name, what do you think?

Reply:
I think that would be better. "Within" a timestamp is difficult to parse since a timestamp is instinctively an instant in time.

Reply:
Done, fixed in https://github.com/ros2/message_filters/commit/41d66703a2679d72bfbf85e4c87fb74d1138de4e